src/iothread.c

```diff
@@ -0,0 +1,622 @@
+    UNUSED(ae);
+    UNUSED(mask);
+
+    IOThread *t = ptr;
+
+    /* Handle fd event first. */
+    serverAssert(fd == getReadEventFd(t->pending_clients_notifier));
+    handleEventNotifier(t->pending_clients_notifier);
+
+    pthread_mutex_lock(&t->pending_clients_mutex);
+    listJoin(t->processing_clients, t->pending_clients);
+    pthread_mutex_unlock(&t->pending_clients_mutex);
+    if (listLength(t->processing_clients) == 0) return;
+
+    listIter li;
+    listNode *ln;
+    listRewind(t->processing_clients, &li);
+    while((ln = listNext(&li))) {
+        client *c = listNodeValue(ln);
+        serverAssert(!(c->io_flags & (CLIENT_IO_READ_ENABLED | CLIENT_IO_WRITE_ENABLED)));
+        /* Main thread must handle clients with CLIENT_CLOSE_ASAP flag, since
+         * we only set io_flags when clients in io thread are freed ASAP. */
+        serverAssert(!(c->flags & CLIENT_CLOSE_ASAP));
+
+        /* Link client in IO thread clients list first. */
+        serverAssert(c->io_thread_client_list_node == NULL);
+        listAddNodeTail(t->clients, c);
+        c->io_thread_client_list_node = listLast(t->clients);
+
+        /* The client is asked to close, we just let main thread free it. */
+        if (c->io_flags & CLIENT_IO_CLOSE_ASAP) {
+            enqueuePendingClientsToMainThread(c, 1);
+            continue;
+        }
+
+        /* Enable read and write and reset some flags. */
+        c->io_flags |= CLIENT_IO_READ_ENABLED | CLIENT_IO_WRITE_ENABLED;
+        c->io_flags &= ~CLIENT_IO_PENDING_COMMAND;
+
+        /* Only bind once, we never remove read handler unless freeing client. */
+        if (!connHasEventLoop(c->conn)) {
+            connRebindEventLoop(c->conn, t->el);
+            serverAssert(!connHasReadHandler(c->conn));
+            connSetReadHandler(c->conn, readQueryFromClient);
+        }
+
+        /* If the client has pending replies, write replies to client. */
+        if (clientHasPendingReplies(c)) {
+            writeToClient(c, 0);
+            if (!(c->io_flags & CLIENT_IO_CLOSE_ASAP) && clientHasPendingReplies(c)) {
+                connSetWriteHandler(c->conn, sendReplyToClient);
+            }
+        }
+    }
+    listEmpty(t->processing_clients);
+}
+
+void IOThreadBeforeSleep(struct aeEventLoop *el) {
+    IOThread *t = el->privdata[0];
+
+    /* Handle pending data(typical TLS). */
+    connTypeProcessPendingData(el);
+
+    /* If any connection type(typical TLS) still has pending unread data don't sleep at all. */
+    aeSetDontWait(el, connTypeHasPendingData(el));
+
+    /* Check if i am pausing */
+    int paused;
+    atomicGetWithSync(t->paused, paused);
+    if (paused == IO_THREAD_PAUSING) {
+        atomicSetWithSync(t->paused, IO_THREAD_PAUSED);
+        /* Wait for resuming */
+        while (paused != IO_THREAD_RESUMING) {
+            atomicGetWithSync(t->paused, paused);
+        }
+        atomicSetWithSync(t->paused, IO_THREAD_UNPAUSED);
```

Comment:
I don't have full grasp of the code but seems like it seems like this is a repeated pattern of setting a variable and waiting until some other thread changes. Maybe, it is possible to write a single function for this. Then, we can optimize/fix in a single place. Consider if this is possible or makes any sense.

If this is the bottleneck, I feel like we can use atomic_compare_exchange api to combine some atomic operations and use relaxed atomic order as there is only two threads involved and both operates on the same atomic variable.

Reply:
yes, main thread sets `PAUSING`, io thread sets `PAUSED` to ACK and busy wait, and then main sets `RESUMING`, io thread sets `UNPAUSED` to ACK.

>  it is possible to write a single function for 

yes, i ever considered it, in order to respond to events timely, I want to call it in other places, such as `afterSleep` of event loop, but it may not be safe since we may take some risky actions, so just call it here, as commented above the `pauseIOThread` function in code. But writing a single function still makes sense.

`relax` is not enough, we need to synchronization and ordering  semantics as PR comments

Reply:
> relax is not enough, we need to synchronization and ordering semantics as PR comments

Maybe I'm missing something but if all we want to do is setting/reading an atomic variable, then memory_order_relaxed should be enough. If we want to make a non-atomic variable visible after an atomic operation, yes then we need more than memory_order_relaxed. I just thought this is not the case here, maybe I'm wrong (I just glanced over the code).

Reply:
> In the main thread, we may want to operate data of IO threads, maybe uninstall event handler, access or operate query/output buffer or resize event loop, we need a clean and safe context to do that. We pause IO thread in IOThreadBeforeSleep, do some jobs and then resume it. To avoid thread suspended, we use busy waiting to confirm the target status. Besides we use atomic variable to make sure memory visibility and ordering. 

when IO thread resumed, it must can see the changes the main thread did

> Relaxed operation: there are no synchronization or ordering constraints imposed on other reads or writes, only this operation's atomicity is guaranteed

https://en.cppreference.com/w/c/atomic/memory_order

if we use `relax`, we also should call `atomic_thread_fence`, https://en.cppreference.com/w/c/atomic/atomic_thread_fence, like valkey https://github.com/valkey-io/valkey/blob/3eb8314be6af0777e69f852b65f933dd9186d30b/src/networking.c#L2437, but i would not like to introduce them, since we need to adapt to multiple versions in `atomicvar.h`, and i tried to use `acquire/release` in https://github.com/redis/redis/pull/7707, but didn't find performance improvement, so just use `memory_order_seq_cst` which requires little thought and is error-free.

Reply:
```c
        atomicSet(t->paused, IO_THREAD_PAUSED);
        /* Wait for resuming */
        while (paused != IO_THREAD_RESUMING) {
            atomicGet(t->paused, paused);
        }
        atomicSetWithSync(t->paused, IO_THREAD_UNPAUSED);
```

Ok, I see. For example, probably we can even do this. See first two calls are not memory_order_cst. 

> didn't find performance improvement, so just use memory_order_seq_cst which requires little thought and is error-free.

Okay, I fully agree. It only matters if this part is bottleneck. We can benchmark at the end and if we realize this is the bottleneck, then we can experiment with it. 

Btw, x86 is called "total store order" system which may not benefit a lot from memory_order_relaxed. Potentially, arm can benefit more.

Reply:
interesting idea, we can try it if there is a bottleneck, remembered in my mind